text/0000-unsafe-aliased.md

```diff
@@ -0,0 +1,337 @@
+
+To write this code in a UB-free way, wrap the fields that are *targets* of self-referential pointers in an `UnsafeAliased`:
+
+```rust
+pub struct S {
+    data: UnsafeAliased<i32>, // <!---- here
+    ptr_to_data: *mut i32,
+}
+```
+
+This lets the compiler know that mutable references to `data` might still have aliases, and so optimizations cannot assume that no aliases exist. That's entirely analogous to how `UnsafeCell` lets the compiler know that *shared* references to this field might undergo mutation.
+
+However, what is *not* analogous is that `&mut S`, when handed to safe code *you do not control*, must still be unique pointers!
+This is because of methods like `mem::swap` that can still assume that their two arguments are non-overlapping.
+(`mem::swap` on two `&mut UnsafeAliased<i32>` may soundly assume that they do not alias.)
+In other words, the safety invariant of `&mut S` still requires full ownership of the entire memory range `S` is stored at; for the duration that a function holds on to the borrow, nobody else may read and write this memory.
+But again, this is a *safety invariant*; it only applies to safe code you do not control. You can write your own code handling `&mut S` and as long as that code is careful not to make use of this memory in the wrong way, potential aliasing is fine.
+You can also hand out a `Pin<&mut S>` to external safe code; the `Pin` wrapper imposes exactly the restrictions needed to ensure that this remains sound (e.g., it prevents `mem::swap`).
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+API sketch:
+
+```rust
+/// The type `UnsafeAliased<T>` lets unsafe code violate
+/// the rule that `&mut UnsafeAliased<T>` may never alias anything else.
+///
+/// However, it is still very risky to have two `&mut UnsafeAliased<T>` that alias
+/// each other. For instance, passing those two references to `mem::swap`
+/// would cause UB. The general advice is to still have a single mutable
+/// reference, and then a bunch of raw pointers that alias it.
+/// If you want to pass such a reference to external safe code,
+/// you need to use techniques such as pinning (`Pin<&mut TypeWithUnsafeAliased>`)
+/// which ensure that both the mutable reference and its aliases remain valid.
+///
+/// Further note that this does *not* lift the requirement that shared references
+/// must be read-only! Use `UnsafeCell` for that.
+#[lang = "unsafe_cell_mut"]
+#[repr(transparent)]
+struct UnsafeAliased<T: ?Sized> {
+    value: T,
+}
+
+impl<T: ?Sized> !Send for UnsafeAliased<T> {}
+
+impl<T> UnsafeCell<T> {
+    /// Constructs a new instance of `UnsafeCell` which will wrap the specified
+    /// value.
+    pub fn new(value: T) -> UnsafeAliased<T> {
+        UnsafeAliased { value }
+    }
+
+    pub fn into_inner(self) -> T {
+        self.value
+    }
+
+    /// Get read-write access to the contents of an `UnsafeAliased`.
+    pub fn get_mut(&mut self) -> *mut T {
```

Comment:
Everything from here to the end of the impl should probably not require `T: Sized` if the type itself doesn't, but maybe that's not that important here.